### PR TITLE
feat(dired): add +dirvish option

### LINKED ISSUE #6397: Support dirvish plugin in dired module
### Describe your request

After trying [dirvish](https://github.com/alexluigit/dirvish) briefly, I think its preview experience is better than ranger. Hope to add a `+dirvish` flag in dired module.

### Briefly explain its use-case

Replace ranger

Please visit https://github.com/alexluigit/dirvish

### PR DESCRIPTION
**Pull request by zrquan**
Add `dirvish` package to dired module.

Regarding the main difference with `ranger.el`, please see this [discussion](https://github.com/alexluigit/dirvish/issues/10).

Fixes #6397

### CODE REVIEW

**Comment by numkem** [comment]
This is working great but it seems to be missing shortcuts. Maybe pair them to what ranger uses?

**Comment by zrquan** [comment]
> This is working great but it seems to be missing shortcuts. Maybe pair them to what ranger uses?

Dirvish is a improved version of Dired (not ranger), it respect all your existing Dired keybindings. So I think there should be as few shortcuts as possible here.

**Comment by numkem** [comment]
I meant more about a shortcut to start dirvish.

**Comment by ckruse** [comment]
Dirvish is an enhancement to dired, you start it by starting dired

**Comment by iyefrat** [comment]
dirvish seems to have an icon extension, it should probably be hooked into the `+icons` flag, both with turning it on and not using `dired-all-the-icons`

**Comment by iyefrat** [comment]
same with `dirvish-vc` to mirror the `dired-git-info` behavior, which is on by default in the module

**Comment by iyefrat** [comment]
Also, thanks for submitting this PR! dirvish looks very promising :)

File: modules/emacs/dired/config.el [hunk 1]
```diff
@@ -160,7 +181,8 @@ we have to clean it up ourselves."
 
 
 (use-package! dired-x
-  :unless (featurep! +ranger)
+  :unless (or (featurep! +dirvish)
+              (featurep! +ranger))
```
Comment (hlissner) [inline modules/emacs/dired/config.el:185]: Please replace with:

```elisp
:unless (featurep! +dirvish)
:unless (featurep! +ranger)
```

File: modules/emacs/dired/config.el [hunk 2]
```diff
@@ -136,6 +136,27 @@ we have to clean it up ourselves."
...
+  (map! :map dirvish-mode-map
+        :n "b" #'dirvish-goto-bookmark
+        :n "z" #'dirvish-show-history
+        :n "f" #'dirvish-file-info-menu
+        :n "F" #'dirvish-toggle-fullscreen
+        :n "l" #'dired-find-file
+        :n "h" #'dired-up-directory
+        :localleader
+        "h" #'dired-omit-mode
+        ))
```
Comment (hlissner) [inline modules/emacs/dired/config.el:157]: Please move up hanging parentheses.

```diff
-        "h" #'dired-omit-mode
-        ))
+        "h" #'dired-omit-mode))
```

File: modules/emacs/dired/config.el [hunk 3]
```diff
@@ -136,6 +136,27 @@ we have to clean it up ourselves."
         ranger-hide-cursor nil))
 
 
+(use-package! dirvish
+  :when (featurep! +dirvish)
+  :after dired
+  :init (dirvish-override-dired-mode)
```
Comment (hlissner) [inline modules/emacs/dired/config.el:142]: use-package's :after keyword has [some load-order implications](https://github.com/jwiegley/use-package/issues/829), and is best avoided. Please replace these two lines with:

```diff
-  :after dired
-  :init (dirvish-override-dired-mode)
+  :defer t
+  :init (after! dired (dirvish-override-dired-mode))
```

It's certainly not as elegant, but I plan to replace our internal uses of `use-package` later to address these issues.

**Review by hlissner — CHANGES_REQUESTED** [review]

**Review by hlissner — APPROVED** [review]

**Comment by hlissner** [comment]
LGTM. Thanks for your help!

**Comment by edgar-vincent** [comment]
The author of `dirvish` [advises against using `dired-rsync`](https://github.com/alexluigit/dirvish/blob/main/CUSTOMIZING.org#use-dirvish-yank-instead-of-dired-rsync) and suggests using `dirvish-yank` instead. Should `dired-rsync` be disabled when `+dirvish` is used?

**Comment by zrquan** [comment]
> The author of `dirvish` [advises against using `dired-rsync`](https://github.com/alexluigit/dirvish/blob/main/CUSTOMIZING.org#use-dirvish-yank-instead-of-dired-rsync) and suggests using `dirvish-yank` instead. Should `dired-rsync` be disabled when `+dirvish` is used?

Maybe you can submit a new PR 😉

**Comment by edgar-vincent** [comment]
I will when I have a moment. Thanks again for your PR! I think dirvish is a great match for Doom :)

**Pull request merged**